File: kits/zig/src/wire.zig
//! The wire: the one way each closed type is written, and the refusals a
//! decoder owes. Article V of the constitution is the whole specification.
//!
//! A value's shape is not on the wire — no names, no tags, no lengths but the
//! ones the law names — so every read and every write is driven by the type
//! the blueprint declares. The type is its notation text, and the record
//! blocks that text may name come from the same blueprint.

const std = @import("std");
const notation = @import("notation");

pub const Error = error{Refused};

pub const key_length = 32;
pub const Key = [key_length]u8;

/// The five things a holder holds, as one typed value.
pub const Invitation = struct {
    warden: Key,
    commitment: Key,
    padlock: Key,
    heir: Key,
    heir_secret: Key,
    hints: []const []const u8,
};

/// The four things a stranger holds: the invitation without the voice.
pub const Card = struct {
    warden: Key,
    commitment: Key,
    padlock: Key,
    hints: []const []const u8,
};

/// One value of one closed type. Which variant a value wears is the
/// blueprint's business, not the wire's: a value that does not match the type
/// it is written under is refused.
pub const Value = union(enum) {
    boolean: bool,
    integer: i64,
    text: []const u8,
    bytes: []const u8,
    /// Thirty-two bytes naming nothing.
    b32: Key,
    /// Thirty-two bytes naming a being. Rides identically to `b32`.
    being: Key,
    invitation: Invitation,
    card: Card,
    list: []const Value,
    /// A `T?` that is not there.
    absent,
    /// A `T?` that is there. `T??` nests, and its two absences are two
    /// distinct byte strings.
    present: *const Value,
    /// A record's fields, in the order the blueprint declares them.
    record: []const Value,
};

/// A decoded value and the arena that owns every byte it points at.
pub const Decoded = struct {
    arena: std.heap.ArenaAllocator,
    value: Value,

    pub fn deinit(self: *Decoded) void {
        self.arena.deinit();
    }
};

const Fault = Error || std.mem.Allocator.Error;

// --------------------------------------------------------------- the shape

const Shape = union(enum) {
    closed: Closed,
    list: []const u8,
    optional: []const u8,
    record: notation.Block,
};

const Closed = enum { boolean, integer, text, bytes, b32, being, invitation, card };

fn closedOf(name: []const u8) ?Closed {
    const table = .{
        .{ "bool", Closed.boolean },
        .{ "int", Closed.integer },
        .{ "text", Closed.text },
        .{ "bytes", Closed.bytes },
        .{ "b32", Closed.b32 },
        .{ "being", Closed.being },
        .{ "invitation", Closed.invitation },
        .{ "card", Closed.card },
    };
    inline for (table) |entry| {
        if (std.mem.eql(u8, name, entry[0])) return entry[1];
    }
    return null;
}

/// The two combinators compose freely, and `?` binds outermost: `[int]?` is
/// an optional list, `[int?]` a list of optionals.
fn shapeOf(type_text: []const u8, records: []const notation.Block) Error!Shape {
    if (type_text.len == 0) return Error.Refused;
    if (type_text[type_text.len - 1] == '?') {
        return .{ .optional = type_text[0 .. type_text.len - 1] };
    }
    if (type_text[0] == '[' and type_text[type_text.len - 1] == ']') {
        return .{ .list = type_text[1 .. type_text.len - 1] };
    }
    if (closedOf(type_text)) |c| return .{ .closed = c };
    for (records) |r| {
        if (std.mem.eql(u8, r.name, type_text)) return .{ .record = r };
    }
    return Error.Refused;
}

// -------------------------------------------------------------- the writing

/// Write one value of one type. Nothing else reaches the wire.
pub fn encode(
    gpa: std.mem.Allocator,
    type_text: []const u8,
    records: []const notation.Block,
    value: Value,
) Fault![]u8 {
    var out: std.ArrayList(u8) = .empty;
    errdefer out.deinit(gpa);
    try write(gpa, &out, type_text, records, value);
    return out.toOwnedSlice(gpa);
}

fn putInt(gpa: std.mem.Allocator, out: *std.ArrayList(u8), n: i64) std.mem.Allocator.Error!void {
    var buf: [8]u8 = undefined;
    std.mem.writeInt(i64, &buf, n, .big);
    try out.appendSlice(gpa, &buf);
}

/// Every length and count is written the way an `int` is, and is non-negative
/// by rule — so a length that cannot be said as one is refused.
fn putCount(gpa: std.mem.Allocator, out: *std.ArrayList(u8), n: usize) Fault!void {
    const said = std.math.cast(i64, n) orelse return Error.Refused;
    try putInt(gpa, out, said);
}

fn putHints(gpa: std.mem.Allocator, out: *std.ArrayList(u8), hints: []const []const u8) Fault!void {
    try putCount(gpa, out, hints.len);
    for (hints) |hint| {
        if (!std.unicode.utf8ValidateSlice(hint)) return Error.Refused;
        try putCount(gpa, out, hint.len);
        try out.appendSlice(gpa, hint);
    }
}

fn write(
    gpa: std.mem.Allocator,
    out: *std.ArrayList(u8),
    type_text: []const u8,
    records: []const notation.Block,
    value: Value,
) Fault!void {
    switch (try shapeOf(type_text, records)) {
        .closed => |c| switch (c) {
            .boolean => {
                const v = switch (value) {
                    .boolean => |b| b,
                    else => return Error.Refused,
                };
                try out.append(gpa, if (v) 1 else 0);
            },
            .integer => {
                const v = switch (value) {
                    .integer => |n| n,
                    else => return Error.Refused,
                };
                try putInt(gpa, out, v);
            },
            .text => {
                const v = switch (value) {
                    .text => |t| t,
                    else => return Error.Refused,
                };
                // A length in front and UTF-8 after it; the length counts
                // bytes, not characters.
                if (!std.unicode.utf8ValidateSlice(v)) return Error.Refused;
                try putCount(gpa, out, v.len);
                try out.appendSlice(gpa, v);
            },
            .bytes => {
                const v = switch (value) {
                    .bytes => |b| b,
                    else => return Error.Refused,
                };
                try putCount(gpa, out, v.len);
                try out.appendSlice(gpa, v);
            },
            .b32 => {
                const v = switch (value) {
                    .b32 => |k| k,
                    else => return Error.Refused,
                };
                try out.appendSlice(gpa, &v);
            },
            .being => {
                const v = switch (value) {
                    .being => |k| k,
                    else => return Error.Refused,
                };
                try out.appendSlice(gpa, &v);
            },
            .invitation => {
                const v = switch (value) {
                    .invitation => |i| i,
                    else => return Error.Refused,
                };
                for ([_]Key{ v.warden, v.commitment, v.padlock, v.heir, v.heir_secret }) |k| {
                    try out.appendSlice(gpa, &k);
                }
                try putHints(gpa, out, v.hints);
            },
            .card => {
                const v = switch (value) {
                    .card => |c2| c2,
                    else => return Error.Refused,
                };
                for ([_]Key{ v.warden, v.commitment, v.padlock }) |k| {
                    try out.appendSlice(gpa, &k);
                }
                try putHints(gpa, out, v.hints);
            },
        },
        .list => |inner| {
            const items = switch (value) {
                .list => |l| l,
                else => return Error.Refused,
            };
            try putCount(gpa, out, items.len);
            for (items) |item| try write(gpa, out, inner, records, item);
        },
        .optional => |inner| switch (value) {
            .absent => try out.append(gpa, 0),
            .present => |held| {
                try out.append(gpa, 1);
                try write(gpa, out, inner, records, held.*);
            },
            else => return Error.Refused,
        },
        .record => |block| {
            const fields = switch (value) {
                .record => |f| f,
                else => return Error.Refused,
            };
            // Its fields, in the order the blueprint declares them, and
            // nothing else.
            if (fields.len != block.fields.len) return Error.Refused;
            for (block.fields, fields) |declared, held| {
                try write(gpa, out, declared.answer orelse return Error.Refused, records, held);
            }
        },
    }
}

// -------------------------------------------------------------- the reading

/// Read one value of one type. Bytes left over after a well-formed value has
/// been read are refused, like every other malformation, in silence.
pub fn decode(
    gpa: std.mem.Allocator,
    type_text: []const u8,
    records: []const notation.Block,
    raw: []const u8,
) Fault!Decoded {
    var arena = std.heap.ArenaAllocator.init(gpa);
    errdefer arena.deinit();

    var reader: Reader = .{ .a = arena.allocator(), .records = records, .rest = raw };
    const value = try reader.read(type_text);
    if (reader.rest.len != 0) return Error.Refused;

    return .{ .arena = arena, .value = value };
}

const Reader = struct {
    a: std.mem.Allocator,
    records: []const notation.Block,
    rest: []const u8,

    fn take(self: *Reader, n: usize) Error![]const u8 {
        if (n > self.rest.len) return Error.Refused;
        const out = self.rest[0..n];
        self.rest = self.rest[n..];
        return out;
    }

    fn takeKey(self: *Reader) Error!Key {
        var out: Key = undefined;
        @memcpy(&out, try self.take(key_length));
        return out;
    }

    fn readInt(self: *Reader) Error!i64 {
        const raw = try self.take(8);
        return std.mem.readInt(i64, raw[0..8], .big);
    }

    /// A negative length, a count beyond the bytes that remain, or a size
    /// beyond what the receiver can address is refused. No type encodes to
    /// zero bytes, so a count can never exceed the bytes that are left.
    fn readCount(self: *Reader) Error!usize {
        const said = try self.readInt();
        if (said < 0) return Error.Refused;
        const n = std.math.cast(usize, said) orelse return Error.Refused;
        if (n > self.rest.len) return Error.Refused;
        return n;
    }

    fn readHints(self: *Reader) Fault![]const []const u8 {
        const count = try self.readCount();
        const out = try self.a.alloc([]const u8, count);
        for (out) |*hint| {
            const length = try self.readCount();
            const raw = try self.take(length);
            if (!std.unicode.utf8ValidateSlice(raw)) return Error.Refused;
            hint.* = raw;
        }
        return out;
    }

    fn read(self: *Reader, type_text: []const u8) Fault!Value {
        switch (try shapeOf(type_text, self.records)) {
            .closed => |c| switch (c) {
                .boolean => {
                    const byte = (try self.take(1))[0];
                    // Where this text does not say what something means, it
                    // is refused.
                    if (byte > 1) return Error.Refused;
                    return .{ .boolean = byte == 1 };
                },
                .integer => return .{ .integer = try self.readInt() },
                .text => {
                    const length = try self.readCount();
                    const raw = try self.take(length);
                    if (!std.unicode.utf8ValidateSlice(raw)) return Error.Refused;
                    return .{ .text = raw };
                },
                .bytes => {
                    const length = try self.readCount();
                    return .{ .bytes = try self.take(length) };
                },
                .b32 => return .{ .b32 = try self.takeKey() },
                .being => return .{ .being = try self.takeKey() },
                .invitation => return .{ .invitation = .{
                    .warden = try self.takeKey(),
                    .commitment = try self.takeKey(),
                    .padlock = try self.takeKey(),
                    .heir = try self.takeKey(),
                    .heir_secret = try self.takeKey(),
                    .hints = try self.readHints(),
                } },
                .card => return .{ .card = .{
                    .warden = try self.takeKey(),
                    .commitment = try self.takeKey(),
                    .padlock = try self.takeKey(),
                    .hints = try self.readHints(),
                } },
            },
            .list => |inner| {
                const count = try self.readCount();
                const items = try self.a.alloc(Value, count);
                for (items) |*item| item.* = try self.read(inner);
                return .{ .list = items };
            },
            .optional => |inner| {
                const marker = (try self.take(1))[0];
                if (marker == 0) return .absent;
                if (marker != 1) return Error.Refused;
                const held = try self.a.create(Value);
                held.* = try self.read(inner);
                return .{ .present = held };
            },
            .record => |block| {
                const fields = try self.a.alloc(Value, block.fields.len);
                for (block.fields, fields) |declared, *held| {
                    held.* = try self.read(declared.answer orelse return Error.Refused);
                }
                return .{ .record = fields };
            },
        }
    }
};

// ------------------------------------------------------------ the comparison

/// Two values are the same value when they wear the same variant and say the
/// same thing. Nothing on the wire depends on this; a caller comparing what it
/// wrote with what it read does.
pub fn equal(x: Value, y: Value) bool {
    if (std.meta.activeTag(x) != std.meta.activeTag(y)) return false;
    return switch (x) {
        .boolean => x.boolean == y.boolean,
        .integer => x.integer == y.integer,
        .text => std.mem.eql(u8, x.text, y.text),
        .bytes => std.mem.eql(u8, x.bytes, y.bytes),
        .b32 => std.mem.eql(u8, &x.b32, &y.b32),
        .being => std.mem.eql(u8, &x.being, &y.being),
        .invitation => std.mem.eql(u8, &x.invitation.warden, &y.invitation.warden) and
            std.mem.eql(u8, &x.invitation.commitment, &y.invitation.commitment) and
            std.mem.eql(u8, &x.invitation.padlock, &y.invitation.padlock) and
            std.mem.eql(u8, &x.invitation.heir, &y.invitation.heir) and
            std.mem.eql(u8, &x.invitation.heir_secret, &y.invitation.heir_secret) and
            sameHints(x.invitation.hints, y.invitation.hints),
        .card => std.mem.eql(u8, &x.card.warden, &y.card.warden) and
            std.mem.eql(u8, &x.card.commitment, &y.card.commitment) and
            std.mem.eql(u8, &x.card.padlock, &y.card.padlock) and
            sameHints(x.card.hints, y.card.hints),
        .list => sameAll(x.list, y.list),
        .absent => true,
        .present => equal(x.present.*, y.present.*),
        .record => sameAll(x.record, y.record),
    };
}

fn sameAll(x: []const Value, y: []const Value) bool {
    if (x.len != y.len) return false;
    for (x, y) |a, b| {
        if (!equal(a, b)) return false;
    }
    return true;
}

fn sameHints(x: []const []const u8, y: []const []const u8) bool {
    if (x.len != y.len) return false;
    for (x, y) |a, b| {
        if (!std.mem.eql(u8, a, b)) return false;
    }
    return true;
}

File: kits/js/src/wire.js
import { refuse } from './refusal.js';
import { SCALARS, baseOf } from './notation.js';

const KEY = 32;
const INT_MIN = -(2n ** 63n);
const INT_MAX = 2n ** 63n - 1n;

class Writer {
  constructor() {
    this.parts = [];
  }
  push(bytes) {
    this.parts.push(bytes);
  }
  // Every length and count is written the same way an int is.
  int(value) {
    if (typeof value !== 'bigint') refuse('NOT_AN_INT', String(value));
    if (value < INT_MIN || value > INT_MAX) refuse('INT_OUT_OF_RANGE', String(value));
    const out = new Uint8Array(8);
    new DataView(out.buffer).setBigInt64(0, value, false);
    this.push(out);
  }
  done() {
    const total = this.parts.reduce((n, p) => n + p.length, 0);
    const out = new Uint8Array(total);
    let at = 0;
    for (const part of this.parts) {
      out.set(part, at);
      at += part.length;
    }
    return out;
  }
}

class Reader {
  constructor(bytes) {
    this.bytes = bytes;
    this.at = 0;
  }
  take(n) {
    if (n > this.bytes.length - this.at) refuse('SHORT_INPUT');
    const out = this.bytes.subarray(this.at, this.at + n);
    this.at += n;
    return out;
  }
  int() {
    const slice = this.take(8);
    return new DataView(slice.buffer, slice.byteOffset, 8).getBigInt64(0, false);
  }
  // A length or a count is an int, and an int that cannot describe a stretch
  // of bytes describes nothing.
  size() {
    const value = this.int();
    if (value < 0n) refuse('NEGATIVE_LENGTH', String(value));
    const remaining = BigInt(this.bytes.length - this.at);
    if (value > remaining) refuse('LENGTH_BEYOND_INPUT', String(value));
    return Number(value);
  }
  flag() {
    const byte = this.take(1)[0];
    if (byte !== 0 && byte !== 1) refuse('NOT_A_FLAG', String(byte));
    return byte === 1;
  }
}

const utf8 = new TextEncoder();
// `ignoreBOM: true` keeps a leading byte order mark in the string rather than
// stripping it: inside a text value a mark is ordinary content, and a decoder
// that ate it would hand back a second spelling of what arrived.
const fromUtf8 = new TextDecoder('utf-8', { fatal: true, ignoreBOM: true });

function bytesOf(value, what) {
  if (!(value instanceof Uint8Array)) refuse('NOT_BYTES', what);
  return value;
}

function writeText(writer, value) {
  if (typeof value !== 'string') refuse('NOT_TEXT', String(value));
  // A kit may not write what no kit may read. JavaScript's strings can hold an
  // unpaired surrogate, which is no code point and no UTF-8; `TextEncoder`
  // would quietly write U+FFFD for it, and a repaired text is a second
  // spelling of a value this protocol names by the hash of its bytes.
  if (/\p{Surrogate}/u.test(value)) refuse('NOT_UTF8');
  const encoded = utf8.encode(value);
  writer.int(BigInt(encoded.length));
  writer.push(encoded);
}

function readText(reader) {
  const slice = reader.take(reader.size());
  try {
    return fromUtf8.decode(slice);
  } catch {
    refuse('NOT_UTF8');
  }
}

function writeKey(writer, value, what) {
  const bytes = bytesOf(value, what);
  if (bytes.length !== KEY) refuse('NOT_A_KEY', `${what} is ${bytes.length} bytes`);
  writer.push(bytes);
}

function writeHints(writer, hints) {
  if (!Array.isArray(hints)) refuse('NOT_A_LIST', 'hints');
  writer.int(BigInt(hints.length));
  for (const hint of hints) writeText(writer, hint);
}

function readHints(reader) {
  const count = reader.size();
  const hints = [];
  for (let i = 0; i < count; i += 1) hints.push(readText(reader));
  return hints;
}

function writeInvitation(writer, value) {
  if (value === null || typeof value !== 'object') refuse('NOT_AN_INVITATION');
  writeKey(writer, value.warden, 'warden');
  writeKey(writer, value.commitment, 'commitment');
  writeKey(writer, value.padlock, 'padlock');
  writeKey(writer, value.heirPublic, 'heirPublic');
  writeKey(writer, value.heirSecret, 'heirSecret');
  writeHints(writer, value.hints);
}

function readInvitation(reader) {
  return {
    warden: reader.take(KEY).slice(),
    commitment: reader.take(KEY).slice(),
    padlock: reader.take(KEY).slice(),
    heirPublic: reader.take(KEY).slice(),
    heirSecret: reader.take(KEY).slice(),
    hints: readHints(reader),
  };
}

// A card is an invitation with the keypair struck out, and the fields it keeps
// ride exactly as they ride there.
function writeCard(writer, value) {
  if (value === null || typeof value !== 'object') refuse('NOT_A_CARD');
  writeKey(writer, value.warden, 'warden');
  writeKey(writer, value.commitment, 'commitment');
  writeKey(writer, value.padlock, 'padlock');
  writeHints(writer, value.hints);
}

function readCard(reader) {
  return {
    warden: reader.take(KEY).slice(),
    commitment: reader.take(KEY).slice(),
    padlock: reader.take(KEY).slice(),
    hints: readHints(reader),
  };
}

function writeScalar(writer, base, value, records) {
  switch (base) {
    case 'bool':
      if (typeof value !== 'boolean') refuse('NOT_A_BOOL', String(value));
      writer.push(Uint8Array.of(value ? 1 : 0));
      return;
    case 'int':
      writer.int(value);
      return;
    case 'text':
      writeText(writer, value);
      return;
    case 'bytes': {
      const bytes = bytesOf(value, 'bytes');
      writer.int(BigInt(bytes.length));
      writer.push(bytes);
      return;
    }
    // `b32` is thirty-two bytes with no length in front, because it never has
    // another size; `being` is a `b32` carrying a pk.
    case 'b32':
      writeKey(writer, value, 'b32');
      return;
    case 'being':
      writeKey(writer, value, 'being');
      return;
    case 'invitation':
      writeInvitation(writer, value);
      return;
    case 'card':
      writeCard(writer, value);
      return;
    default:
      writeRecord(writer, records.get(base), value, records);
  }
}

function readScalar(reader, base, records) {
  switch (base) {
    case 'bool':
      return reader.flag();
    case 'int':
      return reader.int();
    case 'text':
      return readText(reader);
    case 'bytes':
      return reader.take(reader.size()).slice();
    case 'b32':
    case 'being':
      return reader.take(KEY).slice();
    case 'invitation':
      return readInvitation(reader);
    case 'card':
      return readCard(reader);
    default:
      return readRecord(reader, records.get(base), records);
  }
}

// A record is its fields, in the order the blueprint declares them, and
// nothing else — no names on the wire.
function writeRecord(writer, record, value, records) {
  if (value === null || typeof value !== 'object' || Array.isArray(value)) {
    refuse('NOT_A_RECORD', record.name);
  }
  for (const field of record.fields) {
    writeValue(writer, field.type, value[field.name], records);
  }
}

function readRecord(reader, record, records) {
  const out = {};
  for (const field of record.fields) {
    out[field.name] = readValue(reader, field.type, records);
  }
  return out;
}

// The combinators compose freely, so the encoding composes with them.
export function writeValue(writer, type, value, records) {
  if (type.optional) {
    const present = value !== undefined && value !== null;
    writer.push(Uint8Array.of(present ? 1 : 0));
    if (present) writeValue(writer, type.optional, value, records);
    return;
  }
  if (type.list) {
    if (!Array.isArray(value)) refuse('NOT_A_LIST', printableType(type));
    writer.int(BigInt(value.length));
    for (const item of value) writeValue(writer, type.list, item, records);
    return;
  }
  writeScalar(writer, type.base, value, records);
}

export function readValue(reader, type, records) {
  if (type.optional) return reader.flag() ? readValue(reader, type.optional, records) : null;
  if (type.list) {
    const count = reader.size();
    const out = [];
    for (let i = 0; i < count; i += 1) out.push(readValue(reader, type.list, records));
    return out;
  }
  return readScalar(reader, type.base, records);
}

function printableType(type) {
  if (type.optional) return `${printableType(type.optional)}?`;
  if (type.list) return `[${printableType(type.list)}]`;
  return type.base;
}

function known(type, records) {
  const base = baseOf(type);
  if (!SCALARS.has(base) && !records.has(base)) refuse('UNKNOWN_TYPE', base);
}

export function recordsOf(blueprint) {
  return new Map((blueprint?.records ?? []).map((r) => [r.name, r]));
}

export function encode(type, value, records = new Map()) {
  known(type, records);
  const writer = new Writer();
  writeValue(writer, type, value, records);
  return writer.done();
}

// A method's blob is its arguments in declared order, each by the notation,
// concatenated — so many arguments ride in one stretch of bytes with nothing
// between them, and the declared order is the whole of what separates them.
export function encodeAll(types, values, records = new Map()) {
  const writer = new Writer();
  for (const [index, type] of types.entries()) {
    known(type, records);
    writeValue(writer, type, values[index], records);
  }
  return writer.done();
}

export function decodeAll(types, bytes, records = new Map()) {
  const reader = new Reader(bytesOf(bytes, 'input'));
  const out = types.map((type) => {
    known(type, records);
    return readValue(reader, type, records);
  });
  if (reader.at !== bytes.length) refuse('TRAILING_BYTES');
  return out;
}

export function decode(type, bytes, records = new Map()) {
  known(type, records);
  const reader = new Reader(bytesOf(bytes, 'input'));
  const value = readValue(reader, type, records);
  if (reader.at !== bytes.length) refuse('TRAILING_BYTES');
  return value;
}
